src/core/Akka.Streams/Dsl/RepeatPrevious.cs

```diff
@@ -0,0 +1,105 @@
+﻿//-----------------------------------------------------------------------
+// <copyright file="RepeatPrevious.cs" company="Akka.NET Project">
+//     Copyright (C) 2013-2022 .NET Foundation <https://github.com/akkadotnet/akka.net>
+// </copyright>
+//-----------------------------------------------------------------------
+
+using System;
+using Akka.Streams.Stage;
+using Akka.Util;
+
+namespace Akka.Streams.Dsl
+{
+    /// <summary>
+    /// Delegate used to customize replacement functions.
+    ///
+    /// This is used for things like calling <see cref="IDisposable.Dispose"/> on the previous element.
+    /// when a new value is emitted.
+    /// </summary>
+    /// <typeparam name="T">The type of element handled by the <see cref="RepeatPrevious{T}"/></typeparam>
+    public delegate void SwapPrevious<T>(T previousValue, T newValue);
+
+    /// <summary>
+    /// Repeats the previous element from upstream until it's replaced by a new value.
+    ///
+    /// This is designed to allow fan-in stages where output from one of the sources is intermittent / infrequent
+    /// and users just want the previous value to be reused.
+    /// </summary>
+    /// <typeparam name="T">The output type.</typeparam>
+    public sealed class RepeatPrevious<T> : GraphStage<FlowShape<T, T>>
+    {
+        private readonly Inlet<T> _in = new Inlet<T>("RepeatPrevious.in");
+        private readonly Outlet<T> _out = new Outlet<T>("RepeatPrevious.out");
+
+        public override FlowShape<T, T> Shape => new FlowShape<T, T>(_in, _out);
+        private readonly SwapPrevious<T> _swapPrevious;
+
+        /// <summary>
+        /// Do nothing by default
+        /// </summary>
+        private static readonly SwapPrevious<T> DefaultSwap = (value, newValue) => { };
+
+        public RepeatPrevious() : this(DefaultSwap)
+        {
+        }
+
+        public RepeatPrevious(SwapPrevious<T> swapPrevious)
+        {
+            _swapPrevious = swapPrevious;
+        }
+
+        protected override GraphStageLogic CreateLogic(Attributes inheritedAttributes) =>
+            new Logic(this, _swapPrevious);
+
+        private sealed class Logic : InAndOutGraphStageLogic
+        {
+            private readonly RepeatPrevious<T> _stage;
+            private Option<T> _last;
+            private readonly SwapPrevious<T> _swapPrevious;
+
+            public Logic(RepeatPrevious<T> stage, SwapPrevious<T> swapPrevious) : base(stage.Shape)
+            {
+                _stage = stage;
+                _swapPrevious = swapPrevious;
+
+                SetHandler(_stage._in, this);
+                SetHandler(_stage._out, this);
+            }
+
+            public override void OnPush()
+            {
+                var next = Grab(_stage._in);
+                if (_last.HasValue)
+                    _swapPrevious(_last.Value, next);
```

Comment:
Swap function is only needed for scenarios where removing the previous value requires some cleanup - i.e. in my original usecase I needed to call `Dispose` on an `HttpClient`

Reply:
My concern here is that there may or may not be some implicit buffering concerns here.

Is there a risk that either:

 1. There are 'extra' instances of the item getting disposed in a buffer in-between stages
 2. This stage is processing input while the downstream stage is trying to use the last value?

My knowledge of akka streams stage runs is not the best, but I know there are cases where internal buffers can lead to surprising results.